compiler/optimizer/LocalOpts.cpp

```diff
@@ -6932,182 +6939,132 @@ int32_t TR_InvariantArgumentPreexistence::perform()
 
       if (argInfo)
          {
-         if (trace())
+         if (enableTrace)
             traceMsg(comp(), "PREX:    Populating parmInfo of inlined method %s from argInfo %p\n", feMethod->signature(trMemory()), argInfo);
 
          TR_PrexArgInfo *argInfo = comp()->getCurrentInlinedCallArgInfo();
-         int32_t index = 0;
-         for (TR::ParameterSymbol *p = parms.getFirst(); p != NULL; p = parms.getNext(), index++)
+         for (TR::ParameterSymbol *p = parms.getFirst(); p != NULL; p = parms.getNext())
             {
+            int32_t index = p->getOrdinal();
             TR_PrexArgument *arg = argInfo->get(index);
             ParmInfo &parmInfo = _parmInfo[index];
-            if (arg)
+            TR_OpaqueClassBlock *clazzFromMethod = parmInfo.getClass();
+
+            if (!parmInfo.isInvariant())
                {
-               if (trace())
-                  traceMsg(comp(), "PREX:      Parm %d is arg %p parmInfo %p\n", index, arg, &parmInfo);
+               // Clear non-invariant arg info
+               argInfo->set(index, NULL);
+               if (enableTrace)
+                  traceMsg(comp(), "PREX:       parm %d is not invariant\n", index);
+               continue;
                }
-            else
+
+            if (parmInfo.hasKnownObjectIndex())
                {
-               if (trace())
+               TR_ASSERT_FATAL(parmInfo.getKnownObjectIndex() == arg->getKnownObjectIndex(), "Prex arg info should match existing arg info");
+               if (enableTrace)
+                  traceMsg(comp(), "PREX:       parm %d is obj.%d\n", index, parmInfo.getKnownObjectIndex());
+               continue;
+               }
+            if (!clazzFromMethod)
+               {
+               if (enableTrace)
+                  traceMsg(comp(), "PREX:       Can't get method class for parm %d, skipping\n", index);
+               continue;
+               }
+
+            if (!arg)
+               {
+               if (enableTrace)
                   traceMsg(comp(), "PREX:      No argInfo for parm %d\n", index);
                continue;
                }
 
+            if (enableTrace)
+               traceMsg(comp(), "PREX:      Parm %d is arg %p parmInfo %p\n", index, arg, &parmInfo);
+
+            if (arg->hasKnownObjectIndex() && !(arg->getClass() && arg->classIsFixed()))
```

Comment:
might need knot && here

Reply:
`hasKnownObjectIndex` returning true implies the table is enabled